src/main/java/com/exactpro/th2/mstore/ErrorCollector.java

```diff
@@ -0,0 +1,185 @@
+    private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder().setNameFormat("error-collecor-%d").build();
+    private final ScheduledFuture<?> drainFuture;
+    private final MessageRouter<EventBatch> eventRouter;
+    private final EventID rootEvent;
+    private final Lock lock = new ReentrantLock();
+    private Map<String, ErrorMetadata> errors = new HashMap<>();
+
+    public ErrorCollector(@NotNull ScheduledExecutorService executor,
+                          @NotNull MessageRouter<EventBatch> eventRouter,
+                          EventID rootEvent,
+                          long period,
+                          TimeUnit unit) {
+        this.eventRouter = requireNonNull(eventRouter, "Event router can't be null");
+        this.rootEvent = requireNonNull(rootEvent, "Root event can't be null");
+        this.drainFuture = requireNonNull(executor, "Executor can't be null")
+                .scheduleAtFixedRate(this::drain, period, period, unit);
+    }
+
+    public ErrorCollector(@NotNull ScheduledExecutorService executor, @NotNull MessageRouter<EventBatch> eventRouter,
+                          @NotNull EventID rootEvent) {
+        this(executor, eventRouter, rootEvent, 1, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Log error and call the {@link #collect(String)}} method
+     * @param error is used as key indetifyer. Avoid put a lot of unique values
+     */
+    public void collect(Logger logger, String error, Throwable cause) {
+        logger.error(error, cause);
+        collect(error);
+    }
+
+    /**
+     * @param error is used as key indetifyer. Avoid put a lot of unique values
+     */
+    public void collect(String error) {
+        lock.lock();
+        try {
+            errors.compute(error, (key, metadata) -> {
+                if (metadata == null) {
+                    return new ErrorMetadata();
+                }
+                metadata.inc();
+                return metadata;
+            });
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        drainFuture.cancel(true);
+        drain();
+    }
+
+    private void drain() {
+        try {
+            Map<String, ErrorMetadata> map = clear();
+            if (map.isEmpty()) { return; }
+
+            eventRouter.sendAll(Event.start()
+                            .name("mstore internal problem(s): " + calulateTotalQty(map.values()))
+                            .type("InternalError")
+                            .status(Status.FAILED)
+                            .bodyData(new BodyData(map))
+                            .toBatchProto(rootEvent));
+
+        } catch (IOException | RuntimeException e) {
+            LOGGER.error("Drain events task failyre", e);
+        }
+    }
+
+    private Map<String, ErrorMetadata> clear() {
+        try {
+            lock.lock();
```

Comment:
The `lock` method should be called outside the try-finally block. Otherwise, if the `lock` call throws an exception we will call `unlock` method anyway